Add testing across mem_formats to ModuleInfos

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69317
* #69402

Differential Revision: [D33285780](https://our.internmc.facebook.com/intern/diff/D33285780)

mikaylagawarecki committed:
[WIP] Add testing across mem_formats to ModuleInfos

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot ciflow rerun

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

jbschlosser commented on test/test_modules.py:
```diff
@@ -514,6 +514,109 @@ def _to_device(obj):
+
+        # Check that at least one Tensor input has dim == n
+        def _check_dims(obj, n):
+            if isinstance(obj, torch.Tensor):
+                return obj.dim() == n
+            elif isinstance(obj, (tuple, list)):
+                return any(_check_dims(o, n) for o in obj)
+            else:
+                return False
+
+        # Called after _check_dims, when we know that >= 1 tensor can be converted to mem_format
+        def _to_mem_format(obj, mem_format):
+            if isinstance(obj, torch.Tensor):
+                d = obj.dim()
+                if ((mem_format == torch.channels_last and d != 4)
+                   or (mem_format == torch.channels_last_3d and d != 5)):
+                    return obj
+                return obj.to(memory_format=mem_format)
+            elif isinstance(obj, (tuple, list)):
+                return type(obj)(_to_mem_format(o, mem_format) for o in obj)
+            elif isinstance(obj, dict):
+                return {key: _to_mem_format(o, mem_format) for key, o in obj.items()}
+            else:
+                return obj
+
+        def _check_out_mem_format(output, input_mem_format, module_mem_format):
+            if isinstance(output, torch.Tensor):
+                d = output.dim()
+                if (d == 4 and ((input_mem_format == torch.channels_last and (not is_conv or cudnn_greater_7_6))
+                                or (module_mem_format == torch.channels_last and is_conv and cudnn_greater_7_6))):
+                    self.assertTrue(output.is_contiguous(memory_format=torch.channels_last))
+                elif (d == 5 and ((input_mem_format == torch.channels_last_3d and (not is_conv or cudnn_greater_8))
+                                  or (module_mem_format == torch.channels_last_3d and is_conv and cudnn_greater_8))):
+                    self.assertTrue(output.is_contiguous(memory_format=torch.channels_last_3d))
+                else:
+                    self.assertTrue(output.is_contiguous())
+            elif isinstance(output, (tuple, list)):
+                [_check_out_mem_format(o, input_mem_format, module_mem_format) for o in output]
+            else:
+                return
```
For these helper functions, it looks like there is a similar pattern of "traverse a data structure containing tensors and do something" - is it possible to refactor these to use `_traverse_obj()` instead?

jbschlosser commented on test/test_modules.py:
```diff
@@ -514,6 +514,109 @@ def _to_device(obj):
+        module_cls = module_info.module_cls
+        module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
+                                                       requires_grad=False)
+        is_conv = 'Conv' in module_cls.__name__
+        cudnn_greater_7_6 = False
+        cudnn_greater_8 = False
+        if ('cuda' in device) and torch.backends.cudnn.is_available():
+            cudnn_greater_7_6 = torch.backends.cudnn.version() >= 7603
+            cudnn_greater_8 = torch.backends.cudnn.version() >= 8005
```
I think it would be better to have this logic specified in the conv ModuleInfo entries somehow - do you think this is possible somehow?

jbschlosser reviewed (commented):
Thanks for the update! It's looking pretty great, and I love that we are getting broader ModuleInfo coverage as well. I added a few comments below for your consideration :)

jbschlosser commented on test/test_modules.py:
```diff
@@ -514,6 +514,109 @@ def _to_device(obj):
+
+                        # === Compare outputs to a reference if one is specified. ===
+                        # TODO: Handle precision
+                        reference_fn = module_input.reference_fn
+                        if reference_fn is not None:
+                            ref_outputs = reference_fn(m, *args, **kwargs)
+                            self.assertEqual(outputs, ref_outputs)
```
As we discussed, it may be better to compare the outputs against the standard contiguous forward() outputs

jbschlosser commented on test/test_modules.py:
```diff
@@ -514,6 +514,109 @@ def _to_device(obj):
+            if module_input.forward_input is None:
+                continue
+
+            channels_last = _check_dims(module_input.forward_input.args, 4)
+            channels_last_3d = _check_dims(module_input.forward_input.args, 5)
```
tiny stylistic nit: `channels_last` -> `supports_channels_last`

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

jbschlosser commented on test/test_modules.py:
```diff
@@ -528,6 +528,99 @@ def check_backward(cpu_output, gpu_output):
+        module_cls = module_info.module_cls
+        module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
+                                                       requires_grad=False)
+        is_conv = 'Conv' in module_cls.__name__
```
Hm no way to push this into the ModuleInfo entry for Conv* somehow?

mikaylagawarecki replied:
my bad, added a new flag `module_memformat_affects_out` to `ModuleInfo` for this, is that ok?

jbschlosser replied:
sure I'm good with a new flag! Thanks :)

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285780).

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"


Differential Revision: [D33285780](https://our.internmc.facebook.com/intern/diff/D33285780)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285780).

mikaylagawarecki committed:
Update on "Add testing across mem_formats to ModuleInfos"


Differential Revision: [D33285780](https://our.internmc.facebook.com/intern/diff/D33285780)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285780).
